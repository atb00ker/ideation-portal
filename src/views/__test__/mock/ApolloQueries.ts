import { MockedResponse } from '@apollo/client/testing';
import { TopicCategory } from '../../enums/TopicCategory';
import { TopicDepartment } from '../../enums/TopicDepartment';
import { TopicStatus } from '../../enums/TopicStatus';
import { GET_EXCEL_QUERY } from '../../graphql/GetExcel';
import { GET_REPORT_QUERY } from '../../graphql/GetReport';
import { GET_TOPIC_COLLECTION } from '../../graphql/GetTopicsCollection';
import { IGetTopicCollectionInput } from '../../interfaces/IGetTopicCollectionInput';

export const MOCK_GET_TOPIC_COLLECTION: MockedResponse = {
  request: {
    query: GET_TOPIC_COLLECTION,
    variables: {
      limit: parseInt(process.env.HOME_RECORDS_PAGE_SIZE || '5'),
      offset: 0,
      searchFilter: '',
      categoryIdList: Object.keys(TopicCategory).map(key => TopicCategory[key].id),
      departmentIdList: Object.keys(TopicDepartment).map(key => TopicDepartment[key].id),
      statusIdList: Object.keys(TopicStatus).map(key => TopicStatus[key].id),
    } as IGetTopicCollectionInput,
  },
  result: {
    data: {
      topics: [
        {
          id: 'testId',
          title: 'testTitle',
          short_description: 'testDescription',
          status: 0,
          category: 0,
          department: 0,
          topics_users_likes_associations_aggregate: {
            aggregate: {
              count: 10,
            },
          },
        },
      ],
      topics_aggregate: {
        aggregate: {
          count: 10,
        },
      },
    },
  },
};

export const MOCK_GET_EXCEL_QUERY: MockedResponse = {
  request: {
    query: GET_EXCEL_QUERY,
  },
  result: {
    data: {
      GetExcel: {
        excel_link: 'reports.csv',
      },
    },
  },
};

export const MOCK_GET_REPORT: MockedResponse = {
  request: {
    query: GET_REPORT_QUERY,
  },
  result: {
    data: {
      reports_by_pk: {
        site_visitors: 100,
      },
      topics_count_by_category: [
        {
          category: 0,
          count: 1,
        },
      ],
      topics_count_by_status: [
        {
          status: 0,
          count: 1,
        },
      ],
      topics_count_by_department: [
        {
          department: 0,
          count: 1,
        },
      ],
      topics_created_by_month: [
        {
          month: 'Jan',
          topics_created: 10,
        },
      ],
    },
  },
};
