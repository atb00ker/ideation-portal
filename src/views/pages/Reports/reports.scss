.chart-height {
  height: 400px;
}
