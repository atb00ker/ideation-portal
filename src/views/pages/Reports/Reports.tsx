import React from 'react';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import { Chart } from 'react-google-charts';

import SectionLoader from '../../components/ContentState/SectionLoader';
import ServerRequestError from '../../components/ContentState/ServerRequestError';
import GetReport from '../../graphql/GetReport';
import { getTopicCategoryById } from '../../enums/TopicCategory';
import { getTopicStatusById, TopicStatus } from '../../enums/TopicStatus';
import { getTopicDepartmentById } from '../../enums/TopicDepartment';
import './reports.scss';

const Reports = () => {
  const { data: reportData, error: reportError, loading: reportLoading } = GetReport();

  if (reportLoading) return <SectionLoader height='500px' width='100%' />;

  if (reportError) {
    console.error(reportError);
    return <ServerRequestError height='500px' imgHeight='250px' width='100%' />;
  }

  return (
    <Container className='mt-4'>
      <Row>
        <Col className='text-center' xs={3}>
          <h1 className='display-1'>{reportData.reports_by_pk.site_visitors}</h1>
          Total Visitors
        </Col>
        <Col className='text-center' xs={3}>
          <h1 className='display-1'>
            {reportData.topics_count_by_status.find((record: any) => record.status === 0)?.count || 0}
          </h1>
          New Ideas
        </Col>
        <Col className='text-center' xs={3}>
          <h1 className='display-1'>
            {!reportData.topics_count_by_status
              ? reportData.topics_count_by_status
                  .filter((record: any) => {
                    // For all statuses, except the first (proposed) and last (completed), get the id.
                    // The id will be used to reduce and find the sum of the count for
                    // all items in-progress.
                    return [...Array(Object.keys(TopicStatus).length - 2).keys()]
                      .map(index => (index = index + 1))
                      .includes(record.status);
                  })
                  ?.reduce(
                    (totalSum: any, currentRecord: any) => {
                      return currentRecord.count + totalSum.count;
                    },
                    { count: 0 },
                  ) || 0
              : 0}
          </h1>
          Tasks in Progress
        </Col>
        <Col className='text-center' xs={3}>
          <h1 className='display-1'>
            {reportData.topics_count_by_status.find(
              (record: any) => record.status === Object.keys(TopicStatus).length - 1,
            )?.count || 0}
          </h1>
          Completed Ideas
        </Col>
      </Row>
      <Row className='mt-4 mb-5'>
        <Col className='text-center' sm={12} md={6}>
          <Chart
            chartType='PieChart'
            className='chart-height'
            loader={<SectionLoader height='100%' width='100%' />}
            data={[
              ['Category Name', 'Number of Records'],
              ...reportData.topics_count_by_category.map((element: any) => [
                getTopicCategoryById(element.category).name,
                element.count,
              ]),
            ]}
            options={{
              title: 'Ideas by Category',
            }}
            rootProps={{ 'data-testid': 'categoryChart' }}
          />
        </Col>
        <Col className='text-center' sm={12} md={6}>
          <Chart
            chartType='PieChart'
            className='chart-height'
            loader={<SectionLoader height='100%' width='100%' />}
            data={[
              ['Department', 'Number of Records'],
              ...reportData.topics_count_by_department.map((element: any) => [
                getTopicDepartmentById(element.department).name,
                element.count,
              ]),
            ]}
            options={{
              title: 'Ideas by Department',
            }}
            rootProps={{ 'data-testid': 'departmentChart' }}
          />
        </Col>
        <Col className='text-center' sm={12} md={6}>
          <Chart
            className='chart-height'
            chartType='PieChart'
            loader={<SectionLoader height='100%' width='100%' />}
            data={[
              ['Status', 'Number of Records'],
              ...reportData.topics_count_by_status.map((element: any) => [
                getTopicStatusById(element.status).name,
                element.count,
              ]),
            ]}
            options={{
              title: 'Ideas by Status',
            }}
            rootProps={{ 'data-testid': 'statusChart' }}
          />
        </Col>
        {!!reportData.topics_created_by_month.length && (
          <Col className='text-center' sm={12} md={6}>
            <Chart
              chartType='BarChart'
              className='chart-height'
              loader={<SectionLoader height='100%' width='100%' />}
              data={[
                ['Month', 'Ideas Created'],
                ...reportData.topics_created_by_month.map((element: any) => [
                  element.month.trim() || '',
                  element.topics_created || 0,
                ]),
              ]}
              options={{
                title: 'Ideas created / month',
                hAxis: {
                  title: 'Ideas Created',
                  minValue: 0,
                },
                vAxis: {
                  title: 'Month',
                },
                legend: 'none',
              }}
              rootProps={{ 'data-testid': 'monthlyIdeasChart' }}
            />
          </Col>
        )}
        {!reportData.topics_created_by_month.length && (
          <Col className='text-center' sm={12} md={6}>
            <Chart
              className='chart-height'
              chartType='PieChart'
              loader={<SectionLoader height='100%' width='100%' />}
              data={[['A', 'B']]}
              options={{ title: 'Ideas create / month' }}
            />
          </Col>
        )}
      </Row>
    </Container>
  );
};

export default Reports;
