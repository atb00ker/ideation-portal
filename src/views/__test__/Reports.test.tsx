/**
 * @jest-environment jsdom
 */

import { MockedProvider as ApolloMockedProvider } from '@apollo/react-testing';
import '@testing-library/jest-dom';
import 'regenerator-runtime/runtime';
import React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup, act, waitFor, RenderResult } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { HashRouter as Router } from 'react-router-dom';
import Reports from '../pages/Reports/Reports';
import { MOCK_GET_EXCEL_QUERY, MOCK_GET_REPORT } from './mock/ApolloQueries';

const _reportsComponent = (
  <ApolloMockedProvider addTypename={false} mocks={[MOCK_GET_REPORT, MOCK_GET_EXCEL_QUERY]}>
    <Router>
      <Reports></Reports>
    </Router>
  </ApolloMockedProvider>
);

describe('Report page tests', () => {
  // beforeEach(() => {});
  afterEach(() => cleanup);

  it('renders without crashing', () => {
    const root = document.createElement('div');
    ReactDOM.render(_reportsComponent, root);
  });

  it('test search and filter section exists and content section exists', async () => {
    let page: RenderResult | any;
    await act(async () => {
      page = await render(_reportsComponent);
    });
    await waitFor(() => expect(page.getByText('Total Visits')).toBeInTheDocument());
    expect(page.getByText('New Ideas')).toBeInTheDocument();
    expect(page.getByText('Tasks in Progress')).toBeInTheDocument();
    expect(page.getByText('Completed Ideas')).toBeInTheDocument();
    expect(page.getByTestId('reports-proposed-topics')).toHaveTextContent("1");
    expect(page.getByTestId('reports-inprogress-topics')).toHaveTextContent("0");
    expect(page.getByTestId('reports-completed-topics')).toHaveTextContent("0");
  });

  it('report snapshot structure', () => {
    const snapTree = renderer.create(_reportsComponent);
    expect(snapTree.toJSON()).toMatchSnapshot();
  });
});
