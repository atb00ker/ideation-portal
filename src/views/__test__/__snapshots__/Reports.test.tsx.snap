// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report page tests report snapshot structure 1`] = `
<div
  className="position-relative"
  style={
    Object {
      "height": "500px",
      "width": "100%",
    }
  }
>
  <div
    className="round-loader-center position-absolute spinner-border"
    role="status"
  >
    <span
      className="visually-hidden"
      data-testid="section-loading"
    >
      Loading...
    </span>
  </div>
</div>
`;
